### ISSUE TITLE
prometheus_exporter does not support Content-Encoding

### ISSUE THREAD

**Issue by jpds** [vectordotdev/vector#19962]
### A note for the community

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Use Cases

Vector's `prometheus_exporter` does not appear to support gzip content-encoding:

```
$ curl -v --compressed http://localhost:9598/metrics
*   Trying [::1]:9598...
* Connected to localhost (::1) port 9598
> GET /metrics HTTP/1.1
> Host: localhost:9598
> User-Agent: curl/8.4.0
> Accept: */*
> Accept-Encoding: deflate, gzip, br, zstd
>
< HTTP/1.1 200 OK
< content-type: text/plain; version=0.0.4
< content-length: 103854
< date: Tue, 27 Feb 2024 20:42:05 GMT
<
# HELP vector_adaptive_concurrency_averaged_rtt adaptive_concurrency_averaged_rtt
...
```

Whereas other Promtheus exporters such as `node_exporter` automagically do this:

```
$ curl -I --compressed -v http://localhost:9100/metrics
*   Trying [::1]:9100...
* Connected to localhost (::1) port 9100
> HEAD /metrics HTTP/1.1
> Host: localhost:9100
> User-Agent: curl/8.4.0
> Accept: */*
> Accept-Encoding: deflate, gzip, br, zstd
>
< HTTP/1.1 200 OK
HTTP/1.1 200 OK
< Content-Encoding: gzip
Content-Encoding: gzip
< Content-Type: text/plain; version=0.0.4; charset=utf-8
Content-Type: text/plain; version=0.0.4; charset=utf-8
< Date: Tue, 27 Feb 2024 20:43:28 GMT
Date: Tue, 27 Feb 2024 20:43:28 GMT
```

I have many metrics on the same key / labels - so would be beneficial to have this compressed at the network-layer. 

### Version

0.34.1

**Comment by jszwedko** [comment]
👍 this would be a good addition. I think it shouldn't be too difficult to add so I marked it as a "good first issue" if there is a motivated contributor.

**Comment by zamazan4ik** [comment]
For implementing the feature `tower-http`'s CompressionLayer can be used but for now it has some flaws in the implementation: https://github.com/tower-rs/tower-http/issues/215

Since implementing a proper `Accept-Encoding` header is not a very trivial task (e.g. see weights functionality) - I suggest at first rewriting the `prometheus_exporter` functionality to axum, and then using CompressionLayer. Not sure, is it possible or not to use `tower-http` with `hyper`-based server.

**Comment by jpds** [comment]
> Since implementing a proper Accept-Encoding header is not a very trivial task (e.g. see weights functionality) - I suggest at first rewriting the prometheus_exporter functionality to axum,

I don't think this will be an issue as Prometheus only supports `gzip`: https://github.com/prometheus/prometheus/blob/d0ecf9b8fcd06e7b3814251e58d9340ddba791a0/scrape/scrape.go#L716-L721

**Comment by zamazan4ik** [comment]
What about other Prometheus-compatible solutions like VictoriaMetrics?

**Comment by jpds** [comment]
Same: https://github.com/VictoriaMetrics/VictoriaMetrics/blob/46fd0ed6938856806489fbbbbead05c99eee64e0/docs/sd_configs.md?plain=1#L1696

**Comment by zamazan4ik** [comment]
Thanks for checking. So if we can guarantee that no one implementation enables something except gzip and we want to track it in the future - yes, we can enable only gzip. I don't see a reason why we want to implement such a limitation to Vector. Are there any arguments against using a `CompressionLayer`-based solution here?

**Comment by jpds** [comment]
> I don't see a reason why we want to implement such a limitation to Vector.

I don't see the point in adding extra compression algorithms when they're never going to be used by the current tooling?

The whole reason I brought it up was because you mentioned there was a limitation in parsing `Accept-Encoding` and so there was the potential need to rewrite everything into `axum` - but that's not an issue as there's only one encoding to parse.

**Comment by zamazan4ik** [comment]
> I don't see the point in adding extra compression algorithms when they're never going to be used by the current tooling?

I don't see a reason to implement such functionality from scratch instead of using a ready-to-use functionality in the library since in this case we need to maintain less code in Vector. Is it possible to use `CompressionLayer` without rewriting to `axum` or not - it's just a guess, I am not sure.
